state/state.go

```diff
@@ -129,16 +133,12 @@ func (state *HelmState) applyDefaultsTo(spec *ReleaseSpec) {
 	}
 }
 
+// SyncRepos will update the given helm releases
 func (state *HelmState) SyncRepos(helm helmexec.Interface) []error {
 	errs := []error{}
 
 	for _, repo := range state.Repositories {
-		url, err := renderTemplateString(repo.URL)
```

Comment:
Could you try to preserve backward-compatibility here? E.g. try to leave legacy templating as-is.

I believe existing template expressions fail due to syntax error around escaped quotes so we need to somehow resolve that…

Reply:
When templating the entire file this will cause it to error
```
      - name: ingress.enabled
        value: "{{ env \"HOME\" }}"
```

And after the entire file is templated no templates will be left for the `renderTemplateString` to function against.

I believe the best course of action is to release a major version and specify migration steps.  It would be nice if helmfile would transition to a semantic versioning schema.

Reply:
Thanks as always for the thoughtful feedback!

Definitely. That's what I meant where we break backward-compatibility in this PR. I'd like to care existing prod helmfile users not to be forced to fix every helmfile.yaml due to backward-imcompatibilities introduced by a helmfile upgrade.

On the other hand, I want to merge PRs like this quickly for our productivity :)

How can we accomplish both?

And yes, semver and a major version
bump for helmfile would be the future, but that's not what we can do immeadiately! Also, it does't necessary mean we can easily drop existing users with short notice.

So, alternatively, can you introduce a apiVersion key to helmfile.yaml so that we can version current and the default to apiVersion to v1alpha1?

Then, the new templating semantics you introduce via this PR could be enabled via `apiVersion: v1beta1`.

Reply:
I'm not prepared to implement that complexity into the helmfile logic.  I would propose we switch our versioning scheme to semantic version and release a 0.12.0 that contains a breaking change and in the change log we document that upgrading requires making modifications.

I have no issue with someone else implementing the versioning you are requesting, but the migration is fairly simple for anyone wanting to use the updated features, but I don't have the bandwidth to do so.

Reply:
@sstarcher Thanks for the clarification and the feedback. Noted. I'll do my best as a maintainer 😉 

Regarding my last suggestion of introducing `apiVersion` to helmfile, it won't work as I expect because theoretically the vanilla helmfile.yml before applying the entire-file templating can be an invalid yaml.

Reply:
@sstarcher @cmeury Would you mind if I introduced a new convention to helmfile that you MUST name your helmfile.yml suffixed with `.tmpl` so that it looks like `helmfile.yml.tmpl` to invoke the new entire-file templating? Slightly off-topic but it would also reflect the fact that the former helmfile.yml is a valid yaml but the new helmfile.yml introduced after this PR is not.

Very sorry for rushing with many questions. I need feedbacks from experienced helm/helmfile users to shape how the next helmfile should look like 🙏

Reply:
Having a different file extensions seems weird to me, I'd rather handle the "versioning" in the application logic. Could we detect the escaped quotes, warn about the upcoming breaking change and revert to the previous functionality?
Introducing a different version of helmfile seems way too "big" for such a minor change. We're not fundamentally altering the hierarchy of keys.

Reply:
> We're not fundamentally altering the hierarchy of keys.

That's true. I'm still wondering though, if the template syntax is a significant part of the whole helmfile.yml syntax which justifies versioning. But maybe your sentiment is that it is still a minor change regardless of my point? (Just trying to sync up with you

Reply:
> Would you mind if I introduced a new convention to helmfile that you MUST name your helmfile.yml suffixed with .tmpl

If the templated file is invalid yaml I think this is a good idea, otherwise it breaks people's expectations of what the `.yml` file extension means.

Reply:
But isn't Helm doing the same thing? Its templates are not valid YAML but they're still using the `.yaml` extension. Irrespective of whether that's a good or bad thing, I would argue we should follow Helm's practice to meet Helm user's expectations as much as possible.

Reply:
Good point!

Reply:
Good point 😄